File: frontend/src/ui/api/review.ts
import { z } from "zod"

import { apiRequest, type ApiRequestExecutionOptions } from "@/ui/api/http"
import { projectApiPath, type ScopedProjectRef } from "@/ui/api/projectScope"
import { isVirtualStudyReviewProjectId } from "@/ui/guideWalkthrough/guideVirtualProjectIds"
import {
  getVirtualStudyReviewRecallPoint,
  searchVirtualStudyReviewRecallPoints,
} from "@/ui/guideWalkthrough/virtualStudyReviewProject"
import { normalizeRichContent, RichContentSchema } from "@/ui/api/richContent"

export const ReviewTaskSchema = z.object({
  projectId: z.string(),
  reviewTaskId: z.string(),
  inputRangeId: z.string(),
  createdAt: z.string(),
  state: z.string(),
  executedAt: z.string().nullable(),
  resultRangeId: z.string().nullable(),
})
export type ReviewTask = z.infer<typeof ReviewTaskSchema>

export const ConvergenceSchema = z.object({
  projectId: z.string(),
  convergenceId: z.string(),
  seedRangeId: z.string(),
  ruleId: z.string(),
  reviewTaskIds: z.array(z.string()),
  state: z.string(),
  roundCount: z.number().int(),
})
export type Convergence = z.infer<typeof ConvergenceSchema>

export const ReviewChainItemSchema = z.object({
  kind: z.enum(["CONVERGENCE", "REVIEW_TASK"]),
  id: z.string(),
})

export const ReviewChainSchema = z.object({
  projectId: z.string(),
  reviewChainId: z.string(),
  headIndex: z.number().int(),
  state: z.string(),
  queue: z.array(ReviewChainItemSchema),
})
export type ReviewChain = z.infer<typeof ReviewChainSchema>

export const ReviewChainBindingSchema = z.object({
  projectId: z.string(),
  reviewChainId: z.string(),
  entryNodeId: z.string(),
  entryNodeTitle: z.string(),
  entryNodeKind: z.enum(["leaf", "container"]),
  learningTaskId: z.string().nullable(),
  learningTaskTitle: z.string().nullable(),
  childCount: z.number().int().nullable(),
  targetLayerIndex: z.number().int(),
})
export type ReviewChainBinding = z.infer<typeof ReviewChainBindingSchema>

export const ReviewItemBindingSchema = z.object({
  projectId: z.string(),
  kind: z.enum(["CONVERGENCE", "REVIEW_CHAIN"]),
  reviewChainId: z.string(),
  convergenceId: z.string().nullable(),
})
export type ReviewItemBinding = z.infer<typeof ReviewItemBindingSchema>

export const RangeSnapshotSchema = z.object({
  projectId: z.string(),
  rangeId: z.string(),
  recallPointIds: z.array(z.string()),
})
export type RangeSnapshot = z.infer<typeof RangeSnapshotSchema>

export const RecallPointSchema = z.object({
  projectId: z.string(),
  recallPointId: z.string(),
  createdAt: z.string(),
  state: z.enum(["ACTIVE", "DELETED"]),
  deletedAt: z.string().nullable(),
  question: RichContentSchema,
  answer: RichContentSchema,
  anchor: z.object({ instanceId: z.string(), position: z.string() }).nullable(),
  references: z.array(z.string()).default([]),
  insights: z.array(RichContentSchema),
})
export type RecallPoint = z.infer<typeof RecallPointSchema>

export const ReviewRecommendationItemSchema = z.object({
  recallPoint: RecallPointSchema,
  reviewRecommendationIndex: z.number(),
  estimatedMemoryStrength: z.number(),
  weightedSuccessRatio: z.number(),
  lastReviewedAt: z.string().nullable(),
  lastReviewResult: z.enum(["CAN_RECALL", "CANNOT_RECALL"]).nullable(),
  reviewCount: z.number().int(),
})
export type ReviewRecommendationItem = z.infer<typeof ReviewRecommendationItemSchema>

export const ReviewRecommendationPageSchema = z.object({
  items: z.array(ReviewRecommendationItemSchema),
  totalCount: z.number().int(),
  offset: z.number().int(),
  limit: z.number().int(),
  nextOffset: z.number().int().nullable(),
})
export type ReviewRecommendationPage = z.infer<typeof ReviewRecommendationPageSchema>

export const RecallPointReviewProjectionSchema = z.object({
  recallPointId: z.string(),
  calculatedAt: z.string(),
  reviewRecommendationIndex: z.number(),
  estimatedMemoryStrength: z.number(),
  weightedSuccessRatio: z.number(),
  forgettingCurveDecayPerDay: z.number(),
  historyWindowSize: z.number().int(),
  lastReviewedAt: z.string().nullable(),
  lastReviewResult: z.enum(["CAN_RECALL", "CANNOT_RECALL"]).nullable(),
  reviewCount: z.number().int(),
  history: z.array(
    z.object({
      reviewTaskId: z.string(),
      occurredAt: z.string(),
      result: z.enum(["CAN_RECALL", "CANNOT_RECALL"]),
    }),
  ),
})
export type RecallPointReviewProjection = z.infer<typeof RecallPointReviewProjectionSchema>

export function getReviewTask(scope: ScopedProjectRef, reviewTaskId: string) {
  return apiRequest({ path: projectApiPath(scope, `/review-tasks/${reviewTaskId}`), responseSchema: ReviewTaskSchema })
}

export function getReviewTaskBinding(scope: ScopedProjectRef, reviewTaskId: string) {
  return apiRequest({
    path: projectApiPath(scope, `/review-tasks/${reviewTaskId}/binding`),
    responseSchema: ReviewItemBindingSchema,
  })
}

export function getConvergence(scope: ScopedProjectRef, convergenceId: string) {
  return apiRequest({ path: projectApiPath(scope, `/convergences/${convergenceId}`), responseSchema: ConvergenceSchema })
}

export function getConvergenceBinding(scope: ScopedProjectRef, convergenceId: string) {
  return apiRequest({
    path: projectApiPath(scope, `/convergences/${convergenceId}/binding`),
    responseSchema: ReviewItemBindingSchema,
  })
}

export function getReviewChain(scope: ScopedProjectRef, reviewChainId: string) {
  return apiRequest({ path: projectApiPath(scope, `/review-chains/${reviewChainId}`), responseSchema: ReviewChainSchema })
}

export function getReviewChainBinding(scope: ScopedProjectRef, reviewChainId: string) {
  return apiRequest({
    path: projectApiPath(scope, `/review-chains/${reviewChainId}/binding`),
    responseSchema: ReviewChainBindingSchema,
  })
}

export function getRangeSnapshot(scope: ScopedProjectRef, rangeId: string) {
  return apiRequest({ path: projectApiPath(scope, `/ranges/${rangeId}`), responseSchema: RangeSnapshotSchema })
}

export function getRecallPoint(scope: ScopedProjectRef, recallPointId: string) {
  if (isVirtualStudyReviewProjectId(scope.scopedProjectId)) {
    return Promise.resolve(getVirtualStudyReviewRecallPoint(recallPointId))
  }
  return apiRequest({ path: projectApiPath(scope, `/recall-points/${recallPointId}`), responseSchema: RecallPointSchema })
}

export function listRecallPoints(scope: ScopedProjectRef) {
  return apiRequest({ path: projectApiPath(scope, "/recall-points"), responseSchema: z.array(RecallPointSchema) })
}

export function searchRecallPoints(
  scope: ScopedProjectRef,
  params?: {
    q?: string
    limit?: number
  },
  options?: ApiRequestExecutionOptions,
) {
  if (isVirtualStudyReviewProjectId(scope.scopedProjectId)) {
    return Promise.resolve(searchVirtualStudyReviewRecallPoints(params?.q ?? "").slice(0, params?.limit ?? 500))
  }
  const search = new URLSearchParams()
  if (params?.q !== undefined && params.q.trim()) search.set("q", params.q.trim())
  if (params?.limit !== undefined) search.set("limit", String(params.limit))
  const query = search.toString()
  return apiRequest({
    path: projectApiPath(scope, `/recall-points/search${query ? `?${query}` : ""}`),
    responseSchema: z.array(RecallPointSchema),
    signal: options?.signal,
    timeoutMs: options?.timeoutMs,
  })
}

export function getRecallPointReviewProjection(scope: ScopedProjectRef, recallPointId: string) {
  return apiRequest({
    path: projectApiPath(scope, `/recall-points/${recallPointId}/review-projection`),
    responseSchema: RecallPointReviewProjectionSchema,
  })
}

export function listReviewRecommendations(scope: ScopedProjectRef, params?: { offset?: number; limit?: number }) {
  const search = new URLSearchParams()
  if (params?.offset !== undefined) search.set("offset", String(params.offset))
  if (params?.limit !== undefined) search.set("limit", String(params.limit))
  const query = search.toString()
  return apiRequest({
    path: projectApiPath(scope, `/review-recommendations${query ? `?${query}` : ""}`),
    responseSchema: ReviewRecommendationPageSchema,
  })
}

export async function listAllReviewRecommendations(scope: ScopedProjectRef, params?: { limit?: number }) {
  const limit = params?.limit ?? 500
  const items: ReviewRecommendationItem[] = []
  let offset = 0

  for (;;) {
    const page = await listReviewRecommendations(scope, { offset, limit })
    items.push(...page.items)
    if (page.nextOffset == null) {
      return { ...page, items, offset: 0, limit }
    }
    offset = page.nextOffset
  }
}

export function editRecallPoint(
  scope: ScopedProjectRef,
  recallPointId: string,
  params: {
    question: z.infer<typeof RichContentSchema>
    answer: z.infer<typeof RichContentSchema>
    anchor: { instanceId: string; position: string } | null
  },
) {
  return apiRequest({
    path: projectApiPath(scope, `/recall-points/${recallPointId}`),
    method: "PUT",
    body: {
      question: RichContentSchema.parse(normalizeRichContent(params.question)),
      answer: RichContentSchema.parse(normalizeRichContent(params.answer)),
      anchor: params.anchor,
    },
    responseSchema: z.null(),
  })
}

export function deleteRecallPoint(scope: ScopedProjectRef, recallPointId: string) {
  return apiRequest({
    path: projectApiPath(scope, `/recall-points/${recallPointId}`),
    method: "DELETE",
    responseSchema: z.null(),
  })
}

export function commitReviewTask(
  scope: ScopedProjectRef,
  reviewTaskId: string,
  params: {
    canRecall: number[]
    appendedInsights?: { recallPointId: string; insight: z.infer<typeof RichContentSchema> }[]
  },
) {
  return apiRequest({
    path: projectApiPath(scope, `/review-tasks/${reviewTaskId}/commit`),
    method: "POST",
    body: {
      canRecall: params.canRecall,
      appendedInsights: params.appendedInsights?.map((item) => ({
        recallPointId: item.recallPointId,
        insight: RichContentSchema.parse(normalizeRichContent(item.insight)),
      })),
    },
    responseSchema: z.null(),
  })
}

File: adapter/routers/review.py
from fastapi import APIRouter, Depends, Query

from adapter.deps import get_api
from adapter.mappers import (
    convergence_to_dto,
    recall_point_review_projection_to_dto,
    range_snapshot_to_dto,
    recall_point_to_dto,
    review_item_binding_to_dto,
    review_chain_binding_to_dto,
    review_chain_to_dto,
    review_task_to_dto,
)
from adapter.scoped_projects import ScopedProject, resolve_scoped_project
from adapter.schemas import CommitReviewTaskRequest, EditRecallPointRequest
from backend.models.enums import ContentBlockKind
from backend.models.errors import PreconditionFailure
from backend.models.learning_task_node import LearningTaskLeaf
from backend.models.recall_point import Anchor
from backend.models.rich_content import ContentBlock, RichContent
from backend.models.types import ConvergenceId, InstanceId, MediaAssetId, RecallPointId, ReviewChainId
from backend.system.api import SystemAPI


router = APIRouter()

def _to_rich_content(blocks) -> RichContent:
    out: list[ContentBlock] = []
    for b in blocks:
        try:
            kind = ContentBlockKind(str(b.kind))
        except Exception:
            raise PreconditionFailure(f"Invalid ContentBlock.kind: {b.kind}")
        if kind == ContentBlockKind.TEXT:
            out.append(ContentBlock(kind=kind, text=b.text))
        elif kind == ContentBlockKind.IMAGE:
            out.append(ContentBlock(kind=kind, asset_id=None if b.assetId is None else MediaAssetId(b.assetId)))
        else:
            raise PreconditionFailure(f"Unknown ContentBlockKind: {b.kind}")
    return tuple(out)


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/queue")
def get_queue(project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    head, ids = api.get_queue(project.internal_project_id)  # type: ignore[arg-type]
    return {"ok": True, "data": {"headId": None if head is None else str(head), "ids": [str(x) for x in ids]}}


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/review-tasks/{reviewTaskId}")
def get_review_task(reviewTaskId: str, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    rt = api.get_review_task(project.internal_project_id, reviewTaskId)  # type: ignore[arg-type]
    return {"ok": True, "data": review_task_to_dto(rt, public_project_id=project.scoped_project_id)}


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/review-tasks/{reviewTaskId}/binding")
def get_review_task_binding(reviewTaskId: str, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    binding = api.get_review_task_binding(project.internal_project_id, reviewTaskId)  # type: ignore[arg-type]
    return {"ok": True, "data": review_item_binding_to_dto(binding, public_project_id=project.scoped_project_id)}


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/convergences/{convergenceId}")
def get_convergence(convergenceId: str, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    item = api.get_convergence(project.internal_project_id, ConvergenceId(convergenceId))  # type: ignore[arg-type]
    return {"ok": True, "data": convergence_to_dto(item, public_project_id=project.scoped_project_id)}


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/convergences/{convergenceId}/binding")
def get_convergence_binding(convergenceId: str, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    binding = api.get_convergence_binding(project.internal_project_id, ConvergenceId(convergenceId))  # type: ignore[arg-type]
    return {"ok": True, "data": review_item_binding_to_dto(binding, public_project_id=project.scoped_project_id)}


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/review-chains/{reviewChainId}")
def get_review_chain(reviewChainId: str, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    chain = api.get_review_chain(project.internal_project_id, ReviewChainId(reviewChainId))  # type: ignore[arg-type]
    return {"ok": True, "data": review_chain_to_dto(chain, public_project_id=project.scoped_project_id)}


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/review-chains/{reviewChainId}/binding")
def get_review_chain_binding(reviewChainId: str, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    reg = api.get_review_chain_entry_registration(project.internal_project_id, ReviewChainId(reviewChainId))  # type: ignore[arg-type]
    entry_node = api.get_learning_task_node(project.internal_project_id, reg.entry_node)  # type: ignore[arg-type]
    task = api.get_learning_task(project.internal_project_id, entry_node.bound_learning_task_id) if isinstance(entry_node, LearningTaskLeaf) else None  # type: ignore[arg-type]
    return {
        "ok": True,
        "data": review_chain_binding_to_dto(
            project_id=project.scoped_project_id,
            reg=reg,
            entry_node=entry_node,
            learning_task=task,
        ),
    }


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/ranges/{rangeId}")
def get_range(rangeId: str, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    snap = api.get_range_snapshot(project.internal_project_id, rangeId)  # type: ignore[arg-type]
    return {"ok": True, "data": range_snapshot_to_dto(snap, public_project_id=project.scoped_project_id)}


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/recall-points")
def list_recall_points(project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    items = [recall_point_to_dto(rp, public_project_id=project.scoped_project_id) for rp in api.list_recall_points(project.internal_project_id)]  # type: ignore[arg-type]
    return {"ok": True, "data": items}


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/recall-points/search")
def search_recall_points(
    q: str | None = Query(default=None, min_length=0, max_length=200),
    limit: int = Query(default=20, ge=1, le=50),
    project: ScopedProject = Depends(resolve_scoped_project),
    api: SystemAPI = Depends(get_api),
) -> dict:
    items = [recall_point_to_dto(rp, public_project_id=project.scoped_project_id) for rp in api.search_recall_points(project.internal_project_id, query=q, limit=limit)]  # type: ignore[arg-type]
    return {"ok": True, "data": items}


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/recall-points/{recallPointId}")
def get_recall_point(recallPointId: str, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    rp = api.get_recall_point(project.internal_project_id, recallPointId)  # type: ignore[arg-type]
    return {"ok": True, "data": recall_point_to_dto(rp, public_project_id=project.scoped_project_id)}


@router.get("/subjects/{subjectId}/projects/{scopedProjectId}/recall-points/{recallPointId}/review-projection")
def get_recall_point_review_projection(recallPointId: str, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    item = api.get_recall_point_review_projection(project.internal_project_id, RecallPointId(recallPointId))  # type: ignore[arg-type]
    return {"ok": True, "data": recall_point_review_projection_to_dto(item)}

@router.put("/subjects/{subjectId}/projects/{scopedProjectId}/recall-points/{recallPointId}")
def edit_recall_point(
    recallPointId: str, req: EditRecallPointRequest, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)
) -> dict:
    anc = None if req.anchor is None else Anchor(instance_id=InstanceId(req.anchor.instanceId), position=req.anchor.position)
    api.edit_recall_point(  # type: ignore[arg-type]
        project.internal_project_id,
        recallPointId,
        _to_rich_content(req.question),
        _to_rich_content(req.answer),
        anc,
    )
    return {"ok": True, "data": None}


@router.delete("/subjects/{subjectId}/projects/{scopedProjectId}/recall-points/{recallPointId}")
def delete_recall_point(recallPointId: str, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)) -> dict:
    api.delete_recall_point(project.internal_project_id, RecallPointId(recallPointId))  # type: ignore[arg-type]
    return {"ok": True, "data": None}


@router.post("/subjects/{subjectId}/projects/{scopedProjectId}/review-tasks/{reviewTaskId}/commit")
def commit_review_task(
    reviewTaskId: str, req: CommitReviewTaskRequest, project: ScopedProject = Depends(resolve_scoped_project), api: SystemAPI = Depends(get_api)
) -> dict:
    appended = None
    if req.appendedInsights is not None:
        appended = tuple((RecallPointId(x.recallPointId), _to_rich_content(x.insight)) for x in req.appendedInsights)

    api.executor_commit_review_task(  # type: ignore[arg-type]
        project.internal_project_id, reviewTaskId, req.canRecall, appended_insights=appended
    )
    return {"ok": True, "data": None}
